### DISCUSSION TITLE
python import code action

### DISCUSSION THREAD

**Discussion by AdrienLemaire** [LazyVim/LazyVim#2427] [Q&A]
HI there,

I've setup lazyvim to work on a python project, and after enabling extras.lang.python, ruff-lsp & pyright, I realized that imports aren't made available in code actions (something I had gotten used to with coc-python before).

Has anyone resolved this issue ? Do you recommend another mason package to get this working ?

**Comment by ArvineKwok** [comment]
could you run :LspInfo and what's the output? Was pyright or ruff-lsp attched to buffer?

**Comment by AdrienLemaire** [comment]
Thanks @ArvineKwok for helping out!

Here's my LspInfo traceback:

```
 Language client log: /home/dori/.local/state/nvim/lsp.log
 Detected filetype:   python
 
 3 client(s) attached to this buffer: 
 
 Client: ruff_lsp (id: 1, bufnr: [11])
 	filetypes:       python
 	autostart:       true
 	root directory:  Running in single file mode.
 	cmd:             /home/dori/.local/share/nvim/mason/bin/ruff-lsp
 
 Client: pyright (id: 2, bufnr: [11])
 	filetypes:       python
 	autostart:       true
 	root directory:  /home/dori/Projects/Work/project/onnx
 	cmd:             /home/dori/.local/share/nvim/mason/bin/pyright-langserver --stdio
 
 Client: copilot (id: 3, bufnr: [11])
 	filetypes:       
 	autostart:       false
 	root directory:  /home/dori/Projects/Work/project
 	cmd:             node /home/dori/.local/share/nvim/lazy/copilot.lua/copilot/index.js
 
 Configured servers list: yamlls, jsonls, html, pyright, lua_ls, tsserver, marksman, tailwindcss, cssls, ruff_lsp, eslint
 ```
 
 I'm using a binding Space+c+v to select the correct python virtualenv on the file, so I have no warning on my imports. But if I remove an import, then `]d` to the error location, then `Space+c+a` to open code actions, I only get
 
 ```
1: Ruff (F821): Disable for this line ruff_lsp                              │
2: Ruff: Organize Imports             ruff_lsp                              │
3: Ruff: Fix All                      ruff_lsp  
```

I understand that ruff doesn't support auto-fix for F821, and that it should be pyright's responsibility. But when I search there, they seemed to say this is pylance's job more than pyright. pylance isn't available for neovim now to my knowledge, so looking for a different formatter that can replace or complement pyright.

**Reply by lkhphuc** [reply]
In case you didn't know, you can go to the first usage of the symbol and use pyright's autocomplete to automatically import when accepting completion.

**Reply by AdrienLemaire** [reply]
It is less convenient because we have to change into insert mode and remove a character from the variable in order to trigger the autocompletion suggestion.

Also, it doesn't work consistently. For example on `CORS(app)`, it doesn't offer the autocompletion to add `from flask_cors import CORS`.

It's still an improvement for my development flow, thanks for the reminder @lkhphuc :)

**Reply by dycw** [reply]
Can you explain a bit more @lkhphuc , I can't seem to understand what this auto import is.

Indeed, https://github.com/microsoft/pyright/issues/251 confirms that this isn't available to `pyright`. Shame, as I've been trying to move off VSCode for a while!

It seems `coc` has this https://github.com/fannheyward/coc-pyright/issues/752 but I'm not sure about setting that up. This is a huge gap for us Python programmers on neovim I think

**Reply by lkhphuc** [reply]
@dycw I mean you can press enter to accept auto-import completion from nvim-cmp like this.
<img width="694" alt="Screenshot 2024-03-24 at 08 07 01" src="https://github.com/LazyVim/LazyVim/assets/12573521/ccb8cce8-fa80-4f56-b053-9f33873911ab">

You can also use basedpyright by setting `g.lazyvim_python_lsp = "basedpyright"` in your options.py https://github.com/DetachHead/basedpyright?tab=readme-ov-file#import-suggestion-code-actions a fork of pyright.

**Comment by shinonomeow** [comment]
Have you found a way to use code action to import?

**Reply by bro69noscope** [reply]
> Have you found a way to use code action to import?

Hi !

[This](https://github.com/LazyVim/LazyVim/discussions/3685#discussioncomment-9791277) allowed me, after hours of scrambling... to finally do it, using lsp native, after having setup coc just for this feature...

[rope_autoimport was the missing compoment](https://github.com/python-lsp/python-lsp-server/blob/develop/docs/autoimport.md).

Idk how many thousands of missing imports I now have to solve using it to make up for the time it took to find this answer 🤷 

Edit: from what I've tested the providing for the import solving comes from pyflakes and/or pyright so one of those has to be enabled (pyflakes is integrated to pylsp)
